### PR TITLE
Fix a bug where you can't read from remote object stores

### PR DESCRIPTION
**Pull request by derekchiang**
Basically the cause was that we were not distinguishing between EOF error and actual errors, so we keep retrying when we get EOF which results in us never returning from reads.

### CODE REVIEW

**Comment by jdoliner** [comment]
All looks reasonable.

Is there a similar problem with not found errors? I.e. if the system tries to read an object that's not there will it get stuck trying to find the same file again and again?

**Comment by derekchiang** [comment]
That's a good point.  We should probably have a whitelist (or blacklist) of errors that we know to retry (or not).  Lemme investigate a bit.

**Comment by derekchiang** [comment]
@jdoliner alright, now we are only retrying on a whitelist of errors.  Wanna take another look?

File: src/server/pkg/obj/obj.go [hunk 1]
```diff
@@ -4,11 +4,33 @@ import (
...
+			storagegateway.ErrorCodeInternalError,
+			storagegateway.ErrorCodeGatewayInternalError,
+		} {
+			if c == err.Code() {
+				return true
+			}
+		}
+	}
+	return false
+}
```
Comment (jdoliner) [inline src/server/pkg/obj/obj.go]: Could we move this logic into the amazon and google implementations. This file shouldn't have anything specific to either of those implementations. I think we want an `IsRetryable` method on the `Client` interface which will dispatch to the right method.
That way when you implement a new obj store you know that you have to implement the method for your own error types.

**Comment by jdoliner** [comment]
@derekchiang just the one change and then I think this is good to go

**Comment by derekchiang** [comment]
@jdoliner another look?  Also can't merge without a decent meme

**Comment by jdoliner** [comment]
The day I ask you to merge a PR w/o a meme is the day I lose all credibility.

[LGTM](http://lgtm.in/i/qzZEIZw1q)

**Comment by jdoliner** [comment]
Looks gif to me

**Pull request merged**